### PR TITLE
[DRAFT] Enable V8 via NAPI

### PR DESCRIPTION
**Pull request by JunielKatarn**
Draft for initial discussion.
After approval, an equivalent PR will be submitted against `main` and then backported.

- Defines a JSIRuntimeHolder for V8 using NAPI.

Pending verification: task scheduling (under direct debugging).

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/8378)

### CODE REVIEW

File: vnext/Desktop.IntegrationTests/DesktopTestRunner.cpp [hunk 1]
```diff
@@ -83,7 +83,7 @@ shared_ptr<ITestInstance> TestRunner::GetInstance(
   devSettings->platformName = "windesktop";
 
   // Set to JSIEngineOverride::Chakra when testing the Chakra.dll JSI runtime.
-  devSettings->jsiEngineOverride = JSIEngineOverride::ChakraCore;
+  devSettings->jsiEngineOverride = JSIEngineOverride::V8Napi;
```
Comment (JunielKatarn) [inline vnext/Desktop.IntegrationTests/DesktopTestRunner.cpp:86]: Ignore. Will be reverted.

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    napi_ext_task_callback taskCb,
+    void *taskData,
+    uint32_t delayMs,
+    napi_finalize finalizeCb,
+    void *finalizeint) {
+
+  NapiJsiV8RuntimeHolder *holder;
+  auto result = napi_get_instance_data(env, (void **)&holder);
+  if (result != napi_status::napi_ok) {
+    //TODO: Signal error
```
Comment (JunielKatarn) [inline vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp:29]: For reviewers.

How to signal an error condition?
ABI-safe methods should not throw exceptions, right?

File: vnext/JSI/Desktop/packages.config [hunk 3]
```diff
@@ -1,5 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="boost" version="1.72.0.0" targetFramework="native" />
```
Comment (tudorms) [inline vnext/JSI/Desktop/packages.config:3]: >boost [](http://example.com/codeflow?start=15&length=5)

I'm curious what this is needed for?

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp (hunk 2 above)
Comment (tudorms) [reply vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp:29]: If it's really exceptional (looks that way), perhaps a VEC (std::abort).

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    return;
+  }
+
+  //TODO: ABI-safe?
```
Comment (tudorms) [inline vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp:33]: >ABI-safe? [](http://example.com/codeflow?start=10&length=9)

I think we want to call taskCb and pass taskData to it instead (casting into a C++ type is indeed ABI-unsafe).

File: vnext/JSI/Desktop/JSI.Desktop.vcxproj [hunk 5]
```diff
@@ -53,13 +53,14 @@
   <PropertyGroup>
     <!-- We need $(ReactNativeWindowsDir)Chakra for ChakraCoreDebugger.h.
     We should remove it from IncludePath once we retire the ChakraExecutor stack. -->
-    <IncludePath>$(ReactNativeWindowsDir);$(ReactNativeDir)\ReactCommon;$(ReactNativeDir)ReactCommon\callinvoker;$(JSI_SourcePath);$(JSI_Source);$(ReactNativeWindowsDir)Chakra;$(ReactNativeWindowsDir)Common;$(ReactNativeWindowsDir)Shared;$(IncludePath)</IncludePath>
+    <IncludePath>$(ReactNativeWindowsDir);$(FollyDir);$(ReactNativeWindowsDir)stubs;$(ReactNativeDir)\ReactCommon;$(ReactNativeDir)ReactCommon\callinvoker;$(JSI_SourcePath);$(JSI_Source);$(ReactNativeWindowsDir)Chakra;$(ReactNativeWindowsDir)Common;$(ReactNativeWindowsDir)Shared;$(IncludePath)</IncludePath>
+    <ExtraWarningsToDisable>$(ExtraWarningsToDisable);4459</ExtraWarningsToDisable>
```
Comment (NikoAri) [inline vnext/JSI/Desktop/JSI.Desktop.vcxproj:57]: >4459 [](http://example.com/codeflow?start=54&length=4)

Could you add comment why this warning is disabled?

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp (hunk 2 above)
Comment (NikoAri) [reply vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp:29]: If we want to avoid crashes, we can do return error code schemes. Eg HRESULT, GetLastError() etc

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      std::unique_ptr<facebook::jsi::PreparedScriptStore> &&preparedScriptStore) noexcept;
+
+ private:
+  static void ScheduleTaskCallback(
+      napi_env env,
+      napi_ext_task_callback taskCb,
+      void *taskData,
+      uint32_t delayMs,
+      napi_finalize finalizeCb,
+      void *finalizeint);
```
Comment (NikoAri) [inline vnext/JSI/Shared/NapiJsiV8RuntimeHolder.h:30]: >finalizeint [](http://example.com/codeflow?start=12&length=11)

(nit) what does this variable mean?

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+   void InitRuntime() noexcept;
```
Comment (NikoAri) [inline vnext/JSI/Shared/NapiJsiV8RuntimeHolder.h:32]: >    [](http://example.com/codeflow?start=0&length=3)

(nit) spacing

**Comment by NikoAri** [comment]
>   Default = 0, // No JSI, will use the legacy ExecutorFactory

(nit) legacyExecutorFactory with ChakraCore

---
Refers to: vnext/Shared/DevSettings.h:28 in 8ca81fc. [](commit_id = 8ca81fc8bc18d4d5015f9ca07857b5c1da50ec5b, deletion_comment = False)

File: vnext/Shared/OInstance.cpp [hunk 8]
```diff
@@ -151,6 +152,9 @@ std::string GetJSBundleFilePath(const std::string &jsBundleBasePath, const std::
 } // namespace
 
 using namespace facebook;
+using namespace Microsoft::JSI;
+
+using std::make_shared;
```
Comment (NikoAri) [inline vnext/Shared/OInstance.cpp:157]: (nit) why is this needed?

File: vnext/Shared/OInstance.cpp [hunk 9]
```diff
@@ -418,6 +422,25 @@ InstanceImpl::InstanceImpl(
...
+            preparedScriptStore = std::make_unique<facebook::react::BasePreparedScriptStoreImpl>(tempPath);
+          }
+
+          m_devSettings->jsiRuntimeHolder = make_shared<NapiJsiV8RuntimeHolder>(
+              m_devSettings, m_jsThread, std::move(scriptStore), std::move(preparedScriptStore));
+
+          break;
+#else
+          assert(false); // V8 is not available in this build, fallthrough
+          [[fallthrough]];
```
Comment (NikoAri) [inline vnext/Shared/OInstance.cpp:441]: >fallthrough [](http://example.com/codeflow?start=12&length=11)

(nit) can we do additional errorCallback, eg m_devSettings->errorCallback("V8Napi is not available")

File: vnext/Shared/OInstance.cpp [hunk 10]
```diff
@@ -418,6 +422,25 @@ InstanceImpl::InstanceImpl(
           m_devSettings->jsiRuntimeHolder =
               std::make_shared<Microsoft::JSI::ChakraRuntimeHolder>(m_devSettings, m_jsThread, nullptr, nullptr);
           break;
+        case JSIEngineOverride::V8Napi: {
+#if defined(USE_V8)
+          std::unique_ptr<facebook::jsi::ScriptStore> scriptStore = nullptr;
```
Comment (NikoAri) [inline vnext/Shared/OInstance.cpp:427]: >= nullptr [](http://example.com/codeflow?start=66&length=9)

(nit) many guidances prefer leaving out such assignments when well defined default is available

File: vnext/Shared/OInstance.cpp [hunk 11]
```diff
@@ -418,6 +422,25 @@ InstanceImpl::InstanceImpl(
...
+          std::unique_ptr<facebook::jsi::PreparedScriptStore> preparedScriptStore = nullptr;
+
+          char tempPath[MAX_PATH];
+          if (GetTempPathA(MAX_PATH, tempPath)) {
```
Comment (NikoAri) [inline vnext/Shared/OInstance.cpp:431]: >GetTempPathA [](http://example.com/codeflow?start=14&length=12)

This will fail on Unicode paths, please use GetTempPathW

File: vnext/Shared/OInstance.cpp [hunk 12]
```diff
@@ -418,6 +422,25 @@ InstanceImpl::InstanceImpl(
...
+            preparedScriptStore = std::make_unique<facebook::react::BasePreparedScriptStoreImpl>(tempPath);
+          }
```
Comment (NikoAri) [inline vnext/Shared/OInstance.cpp:433]: >} [](http://example.com/codeflow?start=10&length=1)

else assert, errorCallback("...")

File: vnext/Shared/OInstance.cpp [hunk 13]
```diff
@@ -418,6 +422,25 @@ InstanceImpl::InstanceImpl(
...
+
+          m_devSettings->jsiRuntimeHolder = make_shared<NapiJsiV8RuntimeHolder>(
+              m_devSettings, m_jsThread, std::move(scriptStore), std::move(preparedScriptStore));
```
Comment (NikoAri) [inline vnext/Shared/OInstance.cpp:436]: >scriptStore [](http://example.com/codeflow?start=51&length=11)

(nit) this is always nullptr, is that intentional?

**Comment by NikoAri** [comment]
>       if (GetTempPathA(MAX_PATH, tempPath)) {

This should use GetTempPathW

---
Refers to: vnext/Shared/OInstance.cpp:455 in 8ca81fc. [](commit_id = 8ca81fc8bc18d4d5015f9ca07857b5c1da50ec5b, deletion_comment = False)

**Review by NikoAri — CHANGES_REQUESTED** [review]
:clock1:

**Comment by NikoAri** [comment]
>       if (PathFileExistsA(fullBundleFilePath.c_str())) {

Not sure where this path is stored, but if it contains Unicode characters, this will fail. All *A(...) Apis should be switched to *W(...)

---
Refers to: vnext/Shared/OInstance.cpp:580 in 8ca81fc. [](commit_id = 8ca81fc8bc18d4d5015f9ca07857b5c1da50ec5b, deletion_comment = False)

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+namespace Microsoft::JSI {
+
+// See napi_ext_schedule_task_callback definition.
+/*static*/ void NapiJsiV8RuntimeHolder::ScheduleTaskCallback(
+    napi_env env,
+    napi_ext_task_callback taskCb,
+    void *taskData,
+    uint32_t delayMs,
+    napi_finalize finalizeCb,
+    void *finalizeint) {
```
Comment (NikoAri) [inline vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp:24]: >finalizeint [](http://example.com/codeflow?start=10&length=11)

This parameter isn't being used?

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+using namespace facebook::react;
+
+using std::shared_ptr;
+using std::unique_ptr;
+
+namespace Microsoft::JSI {
+
+// See napi_ext_schedule_task_callback definition.
+/*static*/ void NapiJsiV8RuntimeHolder::ScheduleTaskCallback(
+    napi_env env,
```
Comment (NikoAri) [inline vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp:19]: >napi_env [](http://example.com/codeflow?start=4&length=8)

(nit) what are these structures? should they be moved? &&?

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+  m_runtime = MakeNapiJsiRuntime(env);
+  m_ownThreadId = std::this_thread::get_id();
+}
+
+#pragma region facebook::jsi::RuntimeHolderLazyInit
+
+shared_ptr<Runtime> NapiJsiV8RuntimeHolder::getRuntime() noexcept /*override*/
+{
+  std::call_once(m_onceFlag, [this]() { InitRuntime(); });
```
Comment (NikoAri) [inline vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp:79]: >call_once [](http://example.com/codeflow?start=7&length=9)

I'm curious if we are overcomplicating things. Aren't we guaranteed to execute functions on Js thread? Why do we need to care about concurrency?

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp (hunk 15 above)
Comment (vmoroz) [reply vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp:19]: This is part of the NAPI. It is a pointer. It cannot be moved.

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp (hunk 4 above)
Comment (vmoroz) [inline vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp:33]: >//TODO: [](http://example.com/codeflow?start=2&length=7)

No, it is not ABI safe.

File: vnext/Shared/DevSettings.h [hunk 17]
```diff
@@ -30,9 +30,10 @@ enum class JSIEngineOverride : int32_t {
   ChakraCore = 2, // Use the JSIExecutorFactory with ChakraCoreRuntime
   Hermes = 3, // Use the JSIExecutorFactory with Hermes
   V8 = 4, // Use the JSIExecutorFactory with V8
-  V8NAPI = 5, // Use the JSIExecutorFactory with V8-NAPI
+  V8Lite = 5, // Use the JSIExecutorFactory with V8-NAPI
```
Comment (vmoroz) [inline vnext/Shared/DevSettings.h:33]: >V8Lite [](http://example.com/codeflow?start=2&length=6)

Why is it needed?

File: vnext/Shared/OInstance.cpp (hunk 11 above)
Comment (asklar) [inline vnext/Shared/OInstance.cpp:431]: don't pass MAX_PATH literally here, pass `ARRAYSIZE(tempPath)` instead

Comment (asklar) [reply vnext/Shared/OInstance.cpp:431]: or better yet `std::size(tempPath)`

File: vnext/Shared/OInstance.cpp [hunk 18]
```diff
@@ -418,6 +422,25 @@ InstanceImpl::InstanceImpl(
           m_devSettings->jsiRuntimeHolder =
               std::make_shared<Microsoft::JSI::ChakraRuntimeHolder>(m_devSettings, m_jsThread, nullptr, nullptr);
           break;
+        case JSIEngineOverride::V8Napi: {
+#if defined(USE_V8)
+          std::unique_ptr<facebook::jsi::ScriptStore> scriptStore = nullptr;
+          std::unique_ptr<facebook::jsi::PreparedScriptStore> preparedScriptStore = nullptr;
+
+          char tempPath[MAX_PATH];
```
Comment (NikoAri) [inline vnext/Shared/OInstance.cpp:430]: >MAX_PATH [](http://example.com/codeflow?start=24&length=8)

(nit) although MAX_PATH isn't huge, it's currently allocated on the stack. It's always better to allocate only required amount. You can view required size by passing 0 into GetTempPath(...), something like:
DWORD requiredSize = GetTempPath(0, nullptr);
string tempPath;
tempPath.reserve(requiredSize);
GetTempPath(tempPath.size(), tempPath.c_str())

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp (hunk 4 above)
Comment (vmoroz) [reply vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp:33]: I was trying to add the correct implementation into this PR, but I do not have permissions.
I have added it into my fork of the same branch:
https://github.com/vmoroz/react-native-windows/commit/a91133d7c8aa8883544c8fe1e523af405094d170

File: vnext/JSI/Desktop/packages.config (hunk 3 above)
Comment (JunielKatarn) [reply vnext/JSI/Desktop/packages.config:3]: Indirect dependencies of the JSI Runtime holder. (Folly, I believe).

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp (hunk 4 above)
Comment (JunielKatarn) [reply vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp]: @vmoroz has solved this via a wrapping class. It has been integrated into this branch.

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.h (hunk 6 above)
Comment (JunielKatarn) [reply vnext/JSI/Shared/NapiJsiV8RuntimeHolder.h]: Finalizing **H**int.

Comment (JunielKatarn) [reply vnext/JSI/Shared/NapiJsiV8RuntimeHolder.h]: @vmoroz How is the hint used? Does it influence the result based on the task/callback state?

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.h (hunk 7 above)
Comment (JunielKatarn) [reply vnext/JSI/Shared/NapiJsiV8RuntimeHolder.h:32]: Formatting will be addressed in the main-facing PR.

File: vnext/Shared/OInstance.cpp (hunk 8 above)
Comment (JunielKatarn) [reply vnext/Shared/OInstance.cpp:157]: Reducing explicit namespace usage and thus line lengths.

File: vnext/Shared/OInstance.cpp (hunk 11 above)
Comment (JunielKatarn) [reply vnext/Shared/OInstance.cpp:434]: Note, the "normal" V8 case also uses `GetTempPathA`.
I'll change this in both.

File: vnext/Shared/OInstance.cpp (hunk 9 above)
Comment (JunielKatarn) [reply vnext/Shared/OInstance.cpp:444]: Because other V8 variants do exactly this, I'd rather add this in a subsequent change.

File: vnext/Shared/OInstance.cpp (hunk 12 above)
Comment (JunielKatarn) [reply vnext/Shared/OInstance.cpp:436]: See `case JSIEngineOverride::V8`.

Keeping behavior consistent with other V8 implementations for this PR.

File: vnext/Shared/OInstance.cpp (hunk 13 above)
Comment (JunielKatarn) [reply vnext/Shared/OInstance.cpp:439]: Fixed. Passing plain `nullptr`.

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp (hunk 16 above)
Comment (JunielKatarn) [reply vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp:106]: I don't know whether is is a guarantee.

The current V8 holder `V8JSIRuntimeHolder` does this validation, so I'm replicating it.

@tudorms did you write this component? Does the thread affinity check make sense?

File: vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp (hunk 4 above)
Comment (JunielKatarn) [reply vnext/JSI/Shared/NapiJsiV8RuntimeHolder.cpp]: Integrated.

File: vnext/Shared/DevSettings.h (hunk 17 above)
Comment (JunielKatarn) [reply vnext/Shared/DevSettings.h:33]: I do not know, but it exists both here and in the Office code base.
Keeping it to make the change less disruptive.

File: vnext/Shared/OInstance.cpp (hunk 18 above)
Comment (JunielKatarn) [reply vnext/Shared/OInstance.cpp:433]: Replicating existing V8 cases as closely as possible.
Will address in a subsequent PR.

**Comment by JunielKatarn** [comment]
> > ```
> >   if (GetTempPathA(MAX_PATH, tempPath)) {
> > ```
> 
> This should use GetTempPathW, other places too. I'm actually surprised that we haven't been bitten by this yet. For example, for if user name contains unicode characters, GetTempPathA will fail, because user name is part of the resulting temp path eg:
> C:\Users[UserName]\AppData\Local\Temp
> 
> Refers to: vnext/Shared/OInstance.cpp:455 in [8ca81fc](https://github.com/microsoft/react-native-windows/commit/8ca81fc8bc18d4d5015f9ca07857b5c1da50ec5b). [](commit_id = [8ca81fc](https://github.com/microsoft/react-native-windows/commit/8ca81fc8bc18d4d5015f9ca07857b5c1da50ec5b), deletion_comment = False)

Can't do without making the change invasive.
Class `BasePreparedScriptStoreImpl` expects `const char*`, and is used by other V8 cases in this method.
Keeping it compatible for now.

**Review by NikoAri — APPROVED** [review]
:shipit:
